### ISSUE TITLE
1 - Introduction: 1.2 Key Components

### ISSUE THREAD

**Issue by JeremyLoh** [JeremyLoh/d2l#3]
https://d2l.ai/chapter_introduction/index.html#key-components

**Supervised learning** - where we try to predict a designated unknown label based on known inputs given a dataset consisting of examples for which the labels are known

Core components, no matter what kind of machine learning problem we take on
1. The **data** that we can learn from
2. A **model** of how to transform the data
3. An **objective function** that quantifies  how well or badly the model is doing
4. An **algorithm** to adjust the model's parameters to optimize the objective function (loss function, the lower the better)

## 1.2.1 Data
To work with data usefully, we typically need to come up with a suitable numerical representation
Each example (or data point, sample) typically consists of a set of attributes called **features** (covariates or inputs), based on which the model must make its predictions

In supervised learning problems, our goal is to predict the value of a special attribute, called the label (or target). It is not part of the model's input

e.g. image data, number indicating the category of photograph (label), RGB of each pixel, data -> 200 x 200 photo will be 200 x 200 x 3 = 120000 numerical values

e.g. Electronic health record data and we want to predict the likelihood that a given patient will survive the next 30 days
Features might consist of a collection of readily available attributes and frequently recorded measurements, e.g. age, vital signs
Label available for training would be a binary value indicating whether each patient in the historical data survived within the 30 day window

In such cases, when every example is characterized by the same number of numerical features, we say that the inputs are **fixed-length vectors** and we call the length of the vectors the **dimensionality** of the data
Fixed length inputs can be convenient, we have one less complication to worry about
Not all data can be represented as fixed-length vectors
e.g. resolution of images mined from internet,  won't be in same resolution or shape
e.g. reviews of text left on website, e.g. amazon. "It stinks!" vs some long winded review that is negative. 

**One major advantage of deep learning over traditional methods is the grace with which modern models can handle varying length data**

The more data we have, the easier our job become
Many of the most exciting models in deep learning do not work without large datasets
We need to have the right data. If we have data with mistakes, or if the chosen features are not predictive of the target quantity of interest, learning will fail
e.g if we have groups of people that are under represented in the training data, it will fail for those groups

## 1.2.2 Models
Most machine learning involves transforming the data in some sense
We might want to build a system that ingests photos and predicts smiley-ness
By model, we denote the computational machinery for ingesting data of one type, and spitting out predictions of a possibly different type

We are interested in statistical models that can be estimated from data
Deep learning is differentiated from classical approaches by the set of powerful models that it focuses on
These models consist of many successive transformations of the data that are chained together top to bottom, thus the name deep learning

## 1.2.3 Objective Functions
We need to have formal mathematical system of learning machines, and have formal measure of how good or bad our models are
In machine learning and optimization more generally, we call these **objective functions**
By convention, we define objective functions so that the lower is better
Because lower is better, these functions are sometimes called **loss functions**

When trying to predict numerical values, the most common loss function is **squared error** (the square of the difference between the prediction and the ground truth target)
For classification, the most common objective is to minimize **error rate** (the fraction of examples on which our predictions disagree with the ground truth)
Some objectives (e.g. squared error) are easy to optimize, while others (e.g. error rate) are difficult to optimize directly, owing to non-differentiability or other complications.
In these cases, it is common to optimize a **surrogate objective**

During optimization, we think of the loss as a function of the model's parameters, and treat the training dataset as a constant
We learn the best values of our model's parameters by minimizing the loss incurred on a set consisting of some number of examples collected for training
However, doing well on the training data does not guarantee that we will do well on unseen data
So we will typically split the available data into two partitions
1. Training dataset (or training set), for the learning model parameters
2. Test dataset (or test set) which is held out for evaluation

At the end of the day, we typically report how our models perform on both partitions
You could think of training performance as analogous to the scores that a student achieves on the practice exams used to prepare for some real final exam. Even if the results are encouraging, that does not guarantee success on the final exam. Over the course of studying, the student might begin to memorize the practice questions, appearing to master the topic but faltering when faced with previously unseen questions on the actual final exam.

**When a model performs well on the training set but fails to generalize to unseen data, we say that is overfitting to the training data**

## 1.2.4 Optimization Algorithms
Once we have got some data source and representation, a model, and a well-defined objective function, we need an algorithm capable of searching for the best possible parameters for minimizing the loss function
Popular optimization algorithms for deep learning are based on an approach called **gradient descent**
- In short, at each step, this method checks to see, for each parameter, which way the training set loss would move if you perturbed that parameter just a small amount. It then updates the parameter in the direction that lowers the loss.
